airflow/www_rbac/views.py

```diff
@@ -309,70 +327,76 @@ def task_stats(self, session=None):
+                       .group_by(DagRun.dag_id)
+                       .subquery('last_dag_run')
+            )
+            RunningDagRun = (
+                session.query(DagRun.dag_id, DagRun.execution_date)
+                       .join(Dag, Dag.dag_id == DagRun.dag_id)
+                       .filter(DagRun.state == State.RUNNING)
+                       .filter(Dag.is_active == True)  # noqa
+                       .subquery('running_dag_run')
+            )
 
-        UnionTI = union_all(LastTI, RunningTI).alias('union_ti')
-        qry = (
-            session.query(UnionTI.c.dag_id, UnionTI.c.state, sqla.func.count())
-                   .group_by(UnionTI.c.dag_id, UnionTI.c.state)
-        )
+            # Select all task_instances from active dag_runs.
+            # If no dag_run is active, return task instances from most recent dag_run.
+            LastTI = (
+                session.query(TI.dag_id.label('dag_id'), TI.state.label('state'))
+                       .join(LastDagRun,
+                             and_(LastDagRun.c.dag_id == TI.dag_id,
+                                  LastDagRun.c.execution_date == TI.execution_date))
+            )
+            RunningTI = (
+                session.query(TI.dag_id.label('dag_id'), TI.state.label('state'))
+                       .join(RunningDagRun,
+                             and_(RunningDagRun.c.dag_id == TI.dag_id,
+                                  RunningDagRun.c.execution_date == TI.execution_date))
+            )
 
-        data = {}
-        for dag_id, state, count in qry:
-            if dag_id not in data:
-                data[dag_id] = {}
-            data[dag_id][state] = count
-        session.commit()
+            UnionTI = union_all(LastTI, RunningTI).alias('union_ti')
+            qry = (
+                session.query(UnionTI.c.dag_id, UnionTI.c.state, sqla.func.count())
+                       .group_by(UnionTI.c.dag_id, UnionTI.c.state)
+            )
 
-        payload = {}
-        for dag in dagbag.dags.values():
-            payload[dag.safe_dag_id] = []
-            for state in State.task_states:
-                try:
-                    count = data[dag.dag_id][state]
-                except Exception:
-                    count = 0
-                d = {
-                    'state': state,
-                    'count': count,
-                    'dag_id': dag.dag_id,
-                    'color': State.color(state)
-                }
-                payload[dag.safe_dag_id].append(d)
+            data = {}
+            for dag_id, state, count in qry:
+                if filter_dag_ids == {'all_dags'} or dag_id in filter_dag_ids:
+                    if dag_id not in data:
+                        data[dag_id] = {}
+                    data[dag_id][state] = count
+            session.commit()
+
+            for dag in dagbag.dags.values():
+                if filter_dag_ids == {'all_dags'} or dag.dag_id in filter_dag_ids:
```

Comment:
It's less efficient to fetch every dags from the db and then filtering them than filtering at query time. Maybe an `apply_dagids_filter(query, dag_ids)` query could be reused throughout though it's a bit tricky since the `.in_` has to be on the model that's queried.